hasher-matcher-actioner/hmalib/lambdas/api/action_rules_api.py

```diff
@@ -0,0 +1,165 @@
+    response.status = 500
+
+
+def get_action_rules_api(hma_config_table: str) -> bottle.Bottle:
+    # The documentation below expects prefix to be '/action-rules/'
+    action_rules_api = bottle.Bottle()
+    HMAConfig.initialize(hma_config_table)
+
+    @action_rules_api.get("/", apply=[jsoninator])
+    def get_action_rules() -> ActionRulesAPIResponse:
+        """
+        Returns all action rules.
+        """
+        error_message = ""
+        action_rules = []
+
+        try:
+            action_rules = ActionRule.get_all()
+            logger.info("action_rules: %s", action_rules)
+        except Exception as e:
+            error_message = "Unexpected error."
+            handle_unexpected_error(e)
+
+        return ActionRulesAPIResponse(error_message, action_rules)
+
+    @action_rules_api.post("/", apply=[jsoninator(ActionRulesAPIRequest)])
+    def create_action_rule(
+        request: ActionRulesAPIRequest,
+    ) -> ActionRulesAPIResponse:
+        """
+        Creates an action rule.
+        """
+        logger.info("request: %s", request)
+        error_message = ""
+
+        try:
+            hmaconfig.create_config(request.action_rule)
+        except ClientError as e:
+            # TODO this test for "already exists" should be moved to a common place
+            if e.response["Error"]["Code"] == "ConditionalCheckFailedException":
+                error_message = f"An action rule with the name '{request.action_rule.name}' already exists."
+                logger.warning(
+                    "Duplicate action rule creation attempted: %s",
+                    e.response["Error"]["Message"],
+                )
+            else:
+                error_message = "Unexpected error."
+                logger.error(
+                    "Unexpected client error: %s", e.response["Error"]["Message"]
+                )
+                logger.exception(e)
+            response.status = 500
+        except Exception as e:
+            error_message = "Unexpected error."
+            handle_unexpected_error(e)
+
+        return ActionRulesAPIResponse(error_message)
+
+    @action_rules_api.put("/<old_name>", apply=[jsoninator(ActionRulesAPIRequest)])
+    def update_action_rule(
+        request: ActionRulesAPIRequest,
+        old_name: str,
+    ) -> ActionRulesAPIResponse:
+        """
+        Updates an action rule.
+        """
+        logger.info("old_name: %s", old_name)
+        logger.info("request: %s", request)
+        error_message = ""
+
+        if ActionRule.exists(request.action_rule.name):
+            try:
+                hmaconfig.update_config(request.action_rule)
+            except Exception as e:
+                error_message = "Unexpected error."
+                handle_unexpected_error(e)
+        elif ActionRule.exists(old_name):
+            try:
+                hmaconfig.create_config(request.action_rule)
+                hmaconfig.delete_config_by_type_and_name("ActionRule", old_name)
```

Comment:
I think you've already brought it up but this feels weird to me... 
The whole need to create delete thing just for a rename makes me feel like we should treat names as ~immutable in terms of the endpoint but add the ability to "copy and rename" on the UI. This would send a request for create and then could provide a modal that asks the user if they want to delete the old copy. That way changing the name is viewed (accurately as having a more significant overhead).

Reply:
I agree the current state is suboptimal. Adding a unique id column to the config table seems to be the right way to go... I added Issue 588 yesterday to track that: https://github.com/facebook/ThreatExchange/issues/588